domain/application/state/charm.go

```diff
@@ -374,24 +373,68 @@ WHERE uuid = $charmID.uuid;
 
 	stmt, err := s.Prepare(query, archivePath, ident)
 	if err != nil {
-		return "", internalerrors.Errorf("failed to prepare query: %w", err)
+		return "", internalerrors.Errorf("preparing query: %w", err)
 	}
 
 	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		if err := tx.Query(ctx, stmt, ident).Get(&archivePath); err != nil {
 			if errors.Is(err, sqlair.ErrNoRows) {
 				return applicationerrors.CharmNotFound
 			}
-			return internalerrors.Errorf("failed to get charm archive path: %w", err)
+			return err
 		}
 		return nil
 	}); err != nil {
-		return "", internalerrors.Errorf("failed to get charm archive path: %w", domain.CoerceError(err))
+		return "", internalerrors.Errorf("getting charm archive path: %w", err)
 	}
 
 	return archivePath.ArchivePath, nil
 }
 
+// GetCharmArchiveMetadata returns the archive storage path and the sha256 hash
+// for the charm using the charm ID.
+// If the charm does not exist, a [errors.CharmNotFound] error is returned.
+func (s *State) GetCharmArchiveMetadata(ctx context.Context, id corecharm.ID) (archivePath string, hash string, err error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", "", internalerrors.Capture(err)
+	}
+
+	var archivePathAndHash charmArchivePathAndHash
+	ident := charmID{UUID: id.String()}
+
+	// NOTE: In theory, we can store multiple hashes of different formats for a
+	// charm. As of writing, we only store the sha256 hash, but in case this changes
+	// filter to specify we return the sha256 hash.
```

Comment:
We can move that comment to the schema, no need to include it here.

Reply:
Not sure if I agree. This comment pertains to our query to retrieve the hash in this function, and justifies why we pick sha256

Reply:
I'm not sure what value it would add in the schema, since it's obviously the case from the schema we currently only record sha256 hash kind

Reply:
Not if you remove the AND clause. If someone adds another hash, there will be so much work to do, that this will be the least of their concerns.

Ensuring the hash type for all locations isn't something we need to enforce at the moment.